lib/lightning/credentials.ex

```diff
@@ -979,4 +1155,459 @@ defmodule Lightning.Credentials do
+      join: requested_client in OauthClient,
+      on: requested_client.id == ^oauth_client_id,
+      where:
+        token.user_id == ^user_id and
+          token_client.client_id == requested_client.client_id and
+          token_client.client_secret == requested_client.client_secret
+    )
+    |> Lightning.Repo.all()
+  end
+
+  defp select_best_matching_token(
+         available_tokens,
+         requested_scopes,
+         mandatory_scopes_string
+       ) do
+    mandatory_scopes = process_scopes(mandatory_scopes_string, ",")
+
+    requested_scope_set = MapSet.new(requested_scopes)
+    mandatory_scope_set = MapSet.new(mandatory_scopes)
+
+    effective_requested_scopes =
+      MapSet.difference(requested_scope_set, mandatory_scope_set)
+
+    if MapSet.size(effective_requested_scopes) == 0 do
+      available_tokens
+      |> Enum.min_by(
+        fn token ->
+          token_scope_set = MapSet.new(token.scopes)
+
+          effective_token_scopes =
+            MapSet.difference(token_scope_set, mandatory_scope_set)
+
+          {MapSet.size(effective_token_scopes),
+           -DateTime.to_unix(token.updated_at)}
+        end,
+        fn -> nil end
+      )
+    else
+      available_tokens
+      |> Enum.filter(fn token ->
+        token_scope_set = MapSet.new(token.scopes)
+
+        effective_token_scopes =
+          MapSet.difference(token_scope_set, mandatory_scope_set)
+
+        non_mandatory_overlap =
+          MapSet.intersection(effective_token_scopes, effective_requested_scopes)
+          |> MapSet.size()
+
+        non_mandatory_overlap > 0
+      end)
+      |> Enum.max_by(
+        fn token ->
+          token_scope_set = MapSet.new(token.scopes)
+
+          effective_token_scopes =
+            MapSet.difference(token_scope_set, mandatory_scope_set)
+
+          matching_scope_count =
+            MapSet.intersection(
+              effective_token_scopes,
+              effective_requested_scopes
+            )
+            |> MapSet.size()
+
+          unrequested_scope_count =
+            MapSet.difference(effective_token_scopes, effective_requested_scopes)
+            |> MapSet.size()
+
+          effective_requested_count = MapSet.size(effective_requested_scopes)
+
+          exact_match? =
+            matching_scope_count == effective_requested_count &&
+              unrequested_scope_count == 0
+
+          last_updated = DateTime.to_unix(token.updated_at)
+
+          {if(exact_match?, do: 1, else: 0), matching_scope_count,
+           -unrequested_scope_count, last_updated}
+        end,
+        fn -> nil end
+      )
+    end
+  end
+
+  @doc """
+  Processes OAuth scopes with a default space delimiter.
+
+  Convenience function that calls `process_scopes/2` with a space delimiter.
+
+  ## Parameters
+    - `input`: OAuth scope string or token to extract scopes from
+
+  ## Returns
+    - List of normalized scope strings
+  """
+  def process_scopes(input) do
```

Comment:
This function name is a bit vague imo, perhaps `normalize_scopes` or `extract_scopes`?